sips/pending/_posts/2017-01-13-binary-compatibility.md

```diff
@@ -0,0 +1,168 @@
+---
+layout: sip
+title: SIP XX - Improving binary compatibility with @binaryCompatible
+disqus: true
+---
+
+__Dmitry Petrashko__
+
+__first submitted 13 January 2017__
+
+## Introduction ##
+
+Scala is a language which evolves fast and thus made a decision to only promise binary compatibility across minor releases.
+At the same time, there is a demand to develop APIs that live longer than a major release cycle of Scala.
+This SIP introduces an annotation `@binaryCompatible` that checks that `what you write is what you get`.
+It will fail compilation in case emitted methods or their signatures 
+are different from those written by users. 
+As long as signatures of methods in source is not changed, `@binaryCompatible` annotated class 
+will be compatible across major version of Scala. 
+
+## Use Cases
+In case there's a need to develop an API that will be used by clients compiled using different major versions of Scala, 
+the current approach is to either develop them in Java or to use best guess to restrict what Scala features should be used.
+There's also a different approach which is used by SBT: instead of publishing a binary `compiler-interface`, sources are published instead 
+that would be locally compiled.
+
+There's also a use-case of defining a class which is supposed to be also used from Java. 
+`@binaryCompatible` will ensure that there are no not-expected methods that would show up in members of a class or an interface.
+
+Dotty currently uses java defined interfaces as public API for IntelliJ in order to ensure binary compatibility. 
+These interfaces can be replaced by `@binaryCompatible` annotated traits to reach the same goal.  
+
+## Design Guidelines
+`@binaryCompatible` is a feature which is supposed to be used by a small subset of the ecosystem to be binary compatible across major versions of Scala.
+Thus this is designed as an advanced feature that is used rarely and thus is intentionally verbose. 
+It's designed to provide strong guarantees, in some cases sacrificing ease of use.
+ 
+The limitations enforced by `@binaryCompatible` are designed to be an overapproximation: 
+instead of permitting a list of features known to be compatible, `@binaryCompatible` enforces a stronger 
+check which is sufficient to promise binary compatibility. 
+
+## Overview ##
+In order for a class or a trait to succeed compilation with the `@binaryCompatible` annotation it has to be:
+  - defined on the top level;
```

Comment:
> Why limit this to just top-level classes? 

1. Allowing to define `@stableABI` inner classes will most likely fail compilation due to second reason, as they will get successors synthesized to access the outer class;
2. there are several subtleties in how outer class can affect compilation of inner class, eg by converting private method to public;
3. use-case of an inner class can be easily handled by defining a top-level supper class and making inner class extend it;

In order to have a more user-friendly behavior than 1 & 2, and given presence of a simple workaround, I've decided to prohibit it.